### ISSUE TITLE
CUDA test failures on Ampere+ GPUs due to cuDNN TF32 Tensor Core precision

### ISSUE THREAD

**Issue by Zhaoxian-Wu** [IBM/aihwkit#766]
## Description

Multiple CUDA tests fail on GPUs with TF32 Tensor Core support (Ampere and newer architectures, sm >= 80). The root cause is a precision mismatch between the **digital reference path** (which uses cuDNN with TF32 enabled by default) and the **analog simulation path** (which uses CUBLAS `sgemm` in full FP32 precision).

Since PyTorch 1.7, `torch.backends.cudnn.allow_tf32` defaults to `True`. cuDNN operations on Ampere+ GPUs may internally use TF32 Tensor Cores (19-bit: 8-bit exponent + 10-bit mantissa) instead of full FP32 (32-bit: 8-bit exponent + 23-bit mantissa) for matrix multiplications. This introduces up to ~1e-3 absolute error per GEMM operation. Meanwhile, the RPU C++ backend calls CUBLAS `sgemm` directly, which respects `torch.backends.cuda.matmul.allow_tf32` (default `False`) and thus runs in full FP32. The tests compare these two paths using tight tolerances (`decimal=4` or `decimal=6`), causing failures.

### Affected tests

- `tests/test_layers_convolution.py` - Conv3d CUDA tests (`test_torch_original_layer`)
- `tests/test_layers_rnn.py` - VanillaRNN, GRU, LSTM CUDA tests (`test_layer_training`, `test_bidir_layer_training`)
- `tests/test_layers_rnn.py` - LSTMCombinedWeight CUDA tests (`test_layer_training`)

### Affected hardware

The behavior depends on both GPU architecture and cuDNN's internal algorithm selection heuristic:

| Architecture | Example GPUs | Conv3d (small) | RNN/LSTM/GRU |
|---|---|---|---|
| Volta / Turing (sm_70/75) | V100, T4, RTX 2080 | pass | pass |
| Ampere (sm_80/86) | A100, RTX 3090 | untested | likely fail |
| Hopper (sm_90) | H100 | **pass** | **fail** |
| Blackwell (sm_100/120) | B200, RTX PRO 6000 | **fail** | **fail** |

Note: Conv3d passes on H100 because cuDNN decides the equivalent GEMM (M=3, K=128, N=630) is too small for Tensor Cores on Hopper, but Blackwell has a lower activation threshold for Tensor Cores. This makes the issue architecture-dependent and not simply a function of TF32 hardware support.

## How to reproduce

```bash
# On any Ampere+ GPU:
python -m pytest tests/test_layers_rnn.py::RNNLayerTest_VanillaRNNCuda_FloatingPoint_NoBias::test_layer_training -v

# On Blackwell GPUs (also):
python -m pytest tests/test_layers_convolution.py::Convolution3dLayerTest_Conv3dCuda_FloatingPoint_NoBias::test_torch_original_layer -v
```

A standalone diagnostic script (no aihwkit dependency) to check if a given GPU is affected:

```python
import math, torch, torch.nn.functional as F

def check(label, a, b):
    d = (a - b).abs().max().item()
    print(f"  {label:42s}  max|diff|={d:.2e}  [{'DIVERGENT' if d>1.5e-4 else 'ok'}]")
    return d

cap = torch.cuda.get_device_capability()
print(f"GPU: {torch.cuda.get_device_name()} (sm_{cap[0]}{cap[1]})")
print(f"cuDNN: {torch.backends.cudnn.version()}, PyTorch: {torch.__version__}")
print(f"cudnn.allow_tf32={torch.backends.cudnn.allow_tf32}")

worst = 0.0
# Conv3d probe
x = torch.randn(3,2,4,5,6, device="cuda")
w = torch.randn(3,2,4,4,4, device="cuda") / math.sqrt(128)
with torch.no_grad():
    y_on = F.conv3d(x, w, padding=2)
    with torch.backends.cudnn.flags(allow_tf32=False):
        y_off = F.conv3d(x, w, padding=2)
worst = max(worst, check("Conv3d  cuDNN(tf32) vs cuDNN(no-tf32)", y_on, y_off))

# RNN probe
rnn = torch.nn.RNN(4, 5, 2, bidirectional=True).cuda()
xr = torch.randn(10, 3, 4, device="cuda")
with torch.no_grad():
    torch.backends.cudnn.allow_tf32 = True;  yr_on = rnn(xr)[0]
    torch.backends.cudnn.allow_tf32 = False; yr_off = rnn(xr)[0]
    torch.backends.cudnn.allow_tf32 = True
worst = max(worst, check("RNN     cuDNN(tf32) vs cuDNN(no-tf32)", yr_on, yr_off))

print(f"\n{'AFFECTED' if worst > 1.5e-4 else 'NOT AFFECTED'}: max divergence = {worst:.2e}")
```

## Expected behavior

All CUDA tests should pass regardless of GPU architecture.

### Option A: Disable TF32 for cuDNN in tests

Add a `conftest.py` or test fixture that sets `torch.backends.cudnn.allow_tf32 = False` before tests run.

**Pros:**
- Minimal code change (one line)
- Tests run in full FP32 on all paths, so existing tight tolerances remain valid
- No risk of masking real bugs by loosening tolerances

**Cons:**
- Tests no longer exercise the default PyTorch behavior (TF32 is the default since PyTorch 1.7)
- Users running aihwkit in production with default settings will have TF32 enabled, so the test environment diverges from real-world usage
- Does not address the underlying fact that analog and digital outputs will differ in production

### Option B: Dynamic tolerance probing

At test session start, run a lightweight probe that measures the actual cuDNN-vs-non-cuDNN (or TF32-vs-FP32) divergence on the current hardware, then derive test tolerances from the measured value.

```python
# Example probe for Conv3d (cached, runs once per session, ~10ms)
@lru_cache
def _probe_cuda_conv3d_tolerance(in_channels, kernel_size, n_trials=10):
    k = kernel_size
    max_diff = 0.0
    with torch.no_grad():
        for _ in range(n_trials):
            x = torch.randn(3, in_channels, k+1, k+2, k+3, device="cuda")
            w = torch.randn(3, in_channels, k, k, k, device="cuda") / math.sqrt(in_channels * k**3)
            y_cudnn = F.conv3d(x, w, padding=k//2)
            with torch.backends.cudnn.flags(enabled=False):
                y_ref = F.conv3d(x, w, padding=k//2)
            max_diff = max(max_diff, (y_cudnn - y_ref).abs().max().item())
    return max_diff
```

**Pros:**
- Hardware-adaptive: tight on V100 (decimal=6), relaxed only where needed on Blackwell (decimal=2-3)
- Tests exercise the real production code path (TF32 enabled)
- Future-proof: automatically adapts to new GPU architectures without code changes

**Cons:**
- More complex implementation (probe functions + tolerance derivation)
- Probe itself uses random data, so measured tolerance is an estimate (mitigated by multiple trials + safety margin)
- Slightly loosened tolerance could theoretically mask a real regression (mitigated by keeping CPU tests at decimal=6)

### Seeking feedback

I lean toward Option B for its future-proofness, possibly combined with Option A as an opt-in debugging switch. But I'd like to hear from the maintainers:

- Is there a preferred precision contract between the analog and digital paths? (i.e., should they be bit-exact in FP32, or is ~1e-3 tolerance acceptable?)
- Should tests reflect the default production environment (TF32 on), or enforce strict FP32 comparison?
- Are there concerns about dynamic probing adding fragility to CI?
- Any other approach I'm missing (e.g., making the RPU backend also use TF32, or pinning cuDNN algorithm selection)?

Happy to submit a PR for whichever direction is preferred.

## Other information

* Pytorch version: 2.8.0+cu128
* Package version: 0.9.2+cuda (development build)
* OS: Ubuntu 24.04
* Python version: 3.10.18
* CUDA toolkit: 12.8
* cuDNN: 9.1.0.02
* Tested GPUs: NVIDIA RTX PRO 6000 Blackwell (sm_120), NVIDIA H100 NVL (sm_90)

**Comment by PabloCarmona** [comment]
Hello @Zhaoxian-Wu! Thanks for letting us know about it, but why are you using version 0.9.2 of the kit instead of the latest one: 1.1.0? Did you try with the latest version, and the same is happening or not?

**Comment by Zhaoxian-Wu** [comment]
Thanks for the heads-up @PabloCarmona ! I was on v0.9.2 because I hadn't pulled the latest release yet. I've now tested on v1.1.0, and the issue still reproduces. It can be reproduced stably using the code provided above.

**Comment by PabloCarmona** [comment]
OK, we are gonna look into it further then. Thanks @Zhaoxian-Wu ! @maljoras @maljoras-sony, could you help us with this and take a look at the issue that is pointed out? Thanks in advance!

**Comment by maljoras-sony** [comment]
Hi @Zhaoxian-Wu , many thanks for raising the issue and the detailed explanation. In my view, options A seems enough. Note that the tests are only for testing whether the custom GEMM implementation of the layers are correct. If default behavior of pytorch now uses a less accurate compute it is rather pytorch's issue and not one of aihwkit's concern. In any case, the analog simulation will use some kind of physical noise in most cases so that the high numerical precision is not really required.  I would go for option A as it is simply testing for correctness of the custom GEMM implementation. One could add some documentation to note that default GEMM is slightly different from the pure digital AIHWKIT when using convolutions. Many thanks again!

**Comment by Zhaoxian-Wu** [comment]
Hi @maljoras-sony, thanks for your comment. I may push a PR for this issue later.
